src/awkward/contents/unionarray.py

```diff
@@ -1412,9 +1412,13 @@ def _to_arrow(self, pyarrow, mask_node, validbytes, length, options):
             else:
                 this_validbytes = None
 
+            this_length = 0
+            if len(this_index) != 0:
+                this_length = this_index.max() + 1
+
             values.append(
                 content._to_arrow(
-                    pyarrow, mask_node, this_validbytes, this_index.shape[0], options
+                    pyarrow, mask_node, this_validbytes, this_length, options
                 )
             )
```

Comment:
Wow! That was wrong!

When propagating the expected `length` through `_to_arrow`, we should note that the length of a UnionArray possibility must be (at least) one more than the largest value in the possibility's index, not the length of the index itself!

Like, if the index values associated with a possibility are `[391, 440, 87]`, the content's length must be at least `441`, so that the index `440` exists. Not 3!

And if the index values associated with a possibility are `[2, 2, 2, 2, 0, 1, 1]`, then we only need the content to have length `3`, not 7!